### ISSUE TITLE
llvm-dwarfdump should return a non-zero exit code when encountering an error.

### ISSUE THREAD

**Issue by JDevlieghere** [llvm/llvm-project#37667]
|  |  |
| --- | --- |
| Bugzilla Link | [38319](https://llvm.org/bz38319) |
| Version | trunk |
| OS | All |
| CC | @adrian-prantl,@dwblaikie,@pogo59,@wolfy1961 |

## Extended Description 
When dwarfdump encounters an error while dumping, it should return a non-zero exit code in addition to printing the error to stdout. We should update the dump method to return a boolean to indicate whether an error was encountered, and use that to determine the exit code.

**Comment by 11happy** [comment]
Hello @Endilll I am Interested to work on this issue can you please assign this to me.
Thank you

**Comment by Endilll** [comment]
@11happy Go ahead!
Just in case, you don't have to ask me or wait until an issue is assigned to you. We don't have a formal process of assigning work.

**Comment by jh7370** [comment]
I just got pinged by the change to this issue. I'm okay with llvm-dwarfdump returning a non-zero exit code when an actual error is encountered (so this comment won't likely impact the work that is done for this bug), but I also feel like in the majority of cases, those errors should be warnings when not in verify mode, or at the very least an error that doesn't result in immediate exiting of the program. This is because early exiting quickly in a tool like llvm-dwarfdump prevents dumping other potentially useful information (we went to a lot of effort to get llvm-readobj to produce warnings, for example, rather than error).

**Comment by dwblaikie** [comment]
> I just got pinged by the change to this issue. I'm okay with llvm-dwarfdump returning a non-zero exit code when an actual error is encountered (so this comment won't likely impact the work that is done for this bug), but I also feel like in the majority of cases, those errors should be warnings when not in verify mode, or at the very least an error that doesn't result in immediate exiting of the program. This is because early exiting quickly in a tool like llvm-dwarfdump prevents dumping other potentially useful information (we went to a lot of effort to get llvm-readobj to produce warnings, for example, rather than error).

I think the verify mode is too expensive to turn it on as often as we'd like - continuing after an error is an option/we should do that whenever we can. So we could leave these issues as errors, but recoverable errors, and still exit non-zero.

If, for some reason, they need to be warnings, not errors - perhaps we could have a `-Werror` style mode where warnings are promoted to errors for the purposes of styling and exit code. Still just as recoverable.

But I think probably just keeping them as errors (anything we'd like a non-zero exit code on) but making sure we recover as much as possible, would be OK? Does that make sense/suit your use cases, @jh7370 ? Or do you specifically need zero exit code during these sort of lower-grade/recoverable check failures?

**Comment by jh7370** [comment]
> But I think probably just keeping them as errors (anything we'd like a non-zero exit code on) but making sure we recover as much as possible, would be OK? Does that make sense/suit your use cases, @jh7370 ? Or do you specifically need zero exit code during these sort of lower-grade/recoverable check failures?

Recoverable errors are probably my preferred option (i.e. error but continue, and exit with non-zero exit code). I'd add that a non-zero exit code is more important, in my mind, for errors than continuing though (but it's debatable about what is an error versus a warning in the internal code of course).

**Comment by 11happy** [comment]
As I was going through this issue and found the relevant files which would be 'llvm-dwarfdump.cpp` in main of that file I think this check of non zero exit code has already been implemented here : 
```
bool Success = true;
  if (Verify) {
    for (StringRef Object : Objects)
      Success &= handleFile(Object, verifyObjectFile, OutputFile.os());
  } else if (Statistics) {
    for (StringRef Object : Objects)
      Success &= handleFile(Object, collectStatsForObjectFile, OutputFile.os());
  } else if (ShowSectionSizes) {
    for (StringRef Object : Objects)
      Success &= handleFile(Object, collectObjectSectionSizes, OutputFile.os());
  } else if (ShowSources) {
    for (StringRef Object : Objects)
      Success &= handleFile(Object, collectObjectSources, OutputFile.os());
  } else {
    for (StringRef Object : Objects)
      Success &= handleFile(Object, dumpObjectFile, OutputFile.os());
  }

  return Success ? EXIT_SUCCESS : EXIT_FAILURE;
  ```
  further as discussed on discord channel , @TNorthover provided  a commit message which is also relevant in this case:
  https://github.com/llvm/llvm-project/commit/e0fd87cc64d5102f20219b6bad0e337dc4e6b2d2.
  
  If I misunderstood anything please correct me.As I am a new to LLVM community I am open to guidance and suggestions.
  Thank you

**Comment by dwblaikie** [comment]
Yeah, looks like that works. Feel free to manually test a few scenarios and check it's doing the right thing - but can probably  just close this issue out.
Sorry I didn't do so/notice when I committed the fix.

**Comment by 11happy** [comment]
Sure

**Comment by oltolm** [comment]
Fixed by https://github.com/llvm/llvm-project/commit/e0fd87cc64d5102f20219b6bad0e337dc4e6b2d2.
